projects/062 - Sperrin/template-sql/medications.template.sql

```diff
@@ -0,0 +1,132 @@
+--┌─────────────┐
+--│ Medications │
+--└─────────────┘
+
+------------ RESEARCH DATA ENGINEER CHECK ------------
+-- 
+
+-- All prescriptions of certain medications during the study period
+
+-- OUTPUT: Data with the following fields
+-- 	-   PatientId (int)
+--  -   Year
+--  -   MedicationDate
+--	-	MedicationCategory - number of prescriptions for given medication category
+
+--Just want the output, not the messages
+SET NOCOUNT ON;
+
+DECLARE @StartDate datetime;
+DECLARE @EndDate datetime;
+SET @StartDate = '1900-09-01'; --- UPDATE !!!!!!!
+SET @EndDate = '2023-08-31';
+
+--> EXECUTE query-build-rq062-cohort.sql
+
+-- load codesets needed for retrieving medication prescriptions
+
+--> CODESET calcium-channel-blockers:1 beta-blockers:1
+--> CODESET statins:1 ace-inhibitor:1 diuretic:1
+--> CODESET angiotensin-receptor-blockers:1 acetylcholinesterase-inhibitors:1 
+
+-- FIX ISSUE WITH DUPLICATE MEDICATIONS, CAUSED BY SOME CODES APPEARING MULTIPLE TIMES IN #VersionedCodeSets and #VersionedSnomedSets
+
+IF OBJECT_ID('tempdb..#VersionedCodeSets_1') IS NOT NULL DROP TABLE #VersionedCodeSets_1;
+SELECT DISTINCT FK_Reference_Coding_ID, Concept, [Version] INTO #VersionedCodeSets_1 FROM #VersionedCodeSets
+
+IF OBJECT_ID('tempdb..#VersionedSnomedSets_1') IS NOT NULL DROP TABLE #VersionedSnomedSets_1;
+SELECT DISTINCT FK_Reference_SnomedCT_ID, Concept, [Version] INTO #VersionedSnomedSets_1 FROM #VersionedSnomedSets
+
+-- RETRIEVE ALL RELEVANT PRESCRPTIONS FOR THE COHORT
+
+IF OBJECT_ID('tempdb..#medications_rx') IS NOT NULL DROP TABLE #medications_rx;
+SELECT 
+	 m.FK_Patient_Link_ID,
+	CAST(MedicationDate AS DATE) as PrescriptionDate,
+	Concept
+INTO #medications_rx
+FROM SharedCare.GP_Medications m
+INNER JOIN #AllCodes a ON a.code = SuppliedCode
```

Comment:
I updated this bit to make faster. I think the typical syntax we use of `WHERE FK_ID IN (...) OR SNOMED_ID IN (...)` is very slow possibly because it's an either/or. In general when I simplify this to just search by SuppliedCode it's a lot quicker. It's also quicker if I first make a temp table with FK_IDs, then make one with SNOMED_IDs, and then join them together later - see query-build-rq065-cohort-medications.sql, query-build-rq065-cohort-events.sql and query-build-rq065-cohort-events-small.sql for examples of each approach. I think we can use SuppliedCode from now on if we've done the find missing codes thing - because the FK_ID/SNOMED_ID is effectively just doing the same thing.

In this case the "WHERE SuppliedCode IN..." is actually dealt with in the INNER JOIN which means we only get records where the SuppliedCode is also in the #AllCodes table.

I also only selected the columns that are actually used later on, which probably speeds it up a little bit as well.

Reply:
That makes sense - thanks for reducing the runtime from 7+ hours to 4 mins! I've actioned all your points and committed them